src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementHistoriesHandler.php

```diff
@@ -0,0 +1,130 @@
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product\Stock\QueryHandler;
+
+use DateTime;
+use DateTimeImmutable;
+use PrestaShop\PrestaShop\Adapter\Product\Stock\Repository\StockMovementFilter;
+use PrestaShop\PrestaShop\Adapter\Product\Stock\Repository\StockMovementRepository;
+use PrestaShop\PrestaShop\Core\Domain\Product\Stock\QueryResult\RangeStockMovementHistory;
+use PrestaShop\PrestaShop\Core\Domain\Product\Stock\QueryResult\SingleStockMovementHistory;
+use PrestaShop\PrestaShop\Core\Domain\Product\Stock\QueryResult\StockMovement;
+use PrestaShop\PrestaShop\Core\Domain\Product\Stock\QueryResult\StockMovementHistory;
+use PrestaShop\PrestaShop\Core\Domain\Product\Stock\ValueObject\StockId;
+
+abstract class AbstractGetStockMovementHistoriesHandler
+{
+    /**
+     * @var StockMovementRepository
+     */
+    protected $stockMovementRepository;
+
+    public function __construct(StockMovementRepository $stockMovementRepository)
+    {
+        $this->stockMovementRepository = $stockMovementRepository;
+    }
+
+    /**
+     * @return StockMovementHistory[]
+     */
+    protected function createStockMovementHistories(StockId $stockId, int $offset, int $limit): array
+    {
+        $mainFilter = (new StockMovementFilter())->setStockIds($stockId);
+        $rangeFilter = (new StockMovementFilter())->setIsOrder(false);
+
+        return array_map(
+            function (array $historyRow): StockMovementHistory {
+                return $this->createStockMovementHistory($historyRow);
+            },
+            $this->stockMovementRepository->getLastStockMovementHistories(
+                $mainFilter,
+                $rangeFilter,
+                $offset,
+                $limit
+            )
+        );
+    }
+
+    protected function createStockMovementHistory(array $historyRow): StockMovementHistory
+    {
+        if ((int) $historyRow['id_stock_mvt_count'] === 1) {
+            $stockMovementHistory = $this->createSingleStockMovementHistory($historyRow);
+        } else {
+            $stockMovementHistory = $this->createRangeStockMovementHistory($historyRow);
+        }
+
+        return $stockMovementHistory;
+    }
+
+    protected function createSingleStockMovementHistory(array $historyRow): SingleStockMovementHistory
+    {
+        return new SingleStockMovementHistory(
+            new StockMovement(
```

Comment:
`StockMovement` is used by a few functional tests, not only related to movement histories. I assume we should keep this class from business domain, whereas `StockMovementHistory` serves a specific use case more related to statistics overview.
Since each of those 2 types serve distinct purposes, isn't it safer to keep them strictly separated ? If so, a single movement history could require a model from business domain layer.
In the end it only depends on how stricly responsibilities should be splitted:
- `StockMovement` only stores business data
- `StockMovementHistory` only gathers prepared data to display in histories UI
wdyt ?

Reply:
Now query result only accept primitives.